router/grpc.go

```diff
@@ -0,0 +1,308 @@
+			return status.Error(codes.PermissionDenied, "pubkey belongs to someone else")
+		}
+
+		if !bytes.Equal(expectedFeeRecipient.Bytes(), registration.Message.FeeRecipient) {
+			g.Logger.Warn("register_validator called with unexpected fee recipient",
+				zap.String("expected", expectedFeeRecipient.String()),
+				zap.String("got", hex.EncodeToString(registration.Message.FeeRecipient)))
+			return status.Error(codes.PermissionDenied, "incorrect fee recipient")
+		}
+
+	}
+
+	return nil
+}
+
+func (g *guardedServerStream) SendMsg(m interface{}) error {
+	return g.ServerStream.SendMsg(m)
+}
+
+func (g *guardedServerStream) RecvMsg(m interface{}) error {
+	pbMsg, ok := m.(proto.Message)
+	if !ok {
+		g.router.Logger.Warn("Unable to capture proto message from gRPC request")
+		return status.Errorf(codes.Internal, "invalid request")
+	}
+
+	g.router.Logger.Debug("intercepted proto request", zap.String("svc", g.svcName))
+	if err := g.cb(pbMsg, g.nodeAddr); err != nil {
+		return err
+	}
+	return g.ServerStream.RecvMsg(m)
+}
+
+func (g *GRPCRouter) payloadInterceptor() grpc.StreamServerInterceptor {
+	services := map[string]any{
+		"ethereum.eth.v1alpha1.BeaconChain":         struct{}{},
+		"ethereum.eth.v1alpha1.BeaconNodeValidator": struct{}{},
+		"ethereum.eth.v1alpha1.Node":                struct{}{},
+	}
+
+	msgCbs := map[string]validationCb{
+		"PrepareBeaconProposer":        g.validatePrepareBeaconProposer,
+		"SubmitValidatorRegistrations": g.validateRegisterValidators,
+	}
+
+	return func(srv any, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+
+		method := strings.Split(info.FullMethod, "/")
+		_, matched := services[method[1]]
+		if !matched {
+			g.Logger.Warn("unknown service", zap.String("service", method[1]))
+			return status.Errorf(codes.Unimplemented, "unknown service %s", method[1])
+		}
+
+		// Get the authentication header
+		ctx := stream.Context()
+		md, exists := metadata.FromIncomingContext(ctx)
+		if !exists {
+			g.Logger.Warn("no metadata on inbound request", zap.String("service", method[1]))
+			return status.Errorf(codes.Unauthenticated, "no metadata on inbound request")
+		}
+
+		// See https://github.com/prysmaticlabs/prysm/issues/11765
+		var nodeAddr common.Address
+		if method[2] != "DomainData" && method[2] != "SubscribeCommitteeSubnets" {
+			val, exists := md["rprnauth"]
+			if !exists || len(val) < 1 {
+				g.Logger.Debug("grpc access without auth header", zap.String("service", method[1]), zap.String("method", method[2]), zap.Bool("exists", exists))
+				return status.Errorf(codes.Unauthenticated, "headers missing")
+			}
+
+			auth := strings.Split(val[0], ":")
+			if len(auth) != 2 {
+				g.Logger.Debug("grpc access with invalid auth header")
+				return status.Errorf(codes.Unauthenticated, "headers invalid")
+			}
+
+			ac := credentials.AuthenticatedCredential{}
+			ac.Base64URLDecode(auth[0], auth[1])
```

Comment:
The code does not check if `ac.Base64URLDecode()` returns an error. From what I could see (I didn't test it), this should be fine, as `Verify()` will try to marshal a `nil` `Credential` message and this will return an error, failing the verification.

Reply:
weird that golangci isn't catching this :thinking:

Reply:
ah i was holding it wrong.